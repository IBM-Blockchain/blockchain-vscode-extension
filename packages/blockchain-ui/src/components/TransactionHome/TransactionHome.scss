@import '@carbon/layout/scss/_spacing';
@import '@carbon/grid/scss/grid';
@import '@carbon/colors/scss/colors';

body {
    font-size: 14px;
    letter-spacing: 0.01px;
    font-weight: 400;
    -webkit-font-smoothing: antialiased;
    -moz-font-smoothing: antialiased;
    -o-font-smoothing: antialiased;
}

.page-container {
    margin: 0;
    padding: 0;

    h5 {
        padding-bottom: $spacing-06;
    }
}

.inner-container {
    height: 100vh;
    max-width: 100%;
}

.page-contents {
    padding: $spacing-07;
}

.titles-container {
    padding: 0px $spacing-04;

    h2 {
        padding: $spacing-09 0px;
    }
}

.contents-container {
    margin: $spacing-02;
}

