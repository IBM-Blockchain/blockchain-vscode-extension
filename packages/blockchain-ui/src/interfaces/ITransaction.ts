interface ITransaction {
    name: string;
    parameters: Array<{ description: string, name: string, schema: {} }>;
    returns: { type: string };
    tag: Array<string>;
}

export default ITransaction;
