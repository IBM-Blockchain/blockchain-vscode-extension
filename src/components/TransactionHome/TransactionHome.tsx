import React, { Component } from 'react';
import './TransactionHome.scss';
import { Accordion, AccordionItem } from 'carbon-components-react';
import Sidebar from '../TransactionSidebar/TransactionSidebar';

interface PageProps {
    messageData: {smartContracts: Array<string>, activeSmartContract: string};
}

interface PageState {
    smartContracts: string[];
    activeSmartContract: string;
}

class TransactionHome extends Component<PageProps, PageState> {

    constructor(props: Readonly<PageProps>) {
        super(props);
        this.state = {
            smartContracts: this.props.messageData.smartContracts,
            activeSmartContract: this.props.messageData.activeSmartContract
        };
    }

    public switchSmartContract(selectedSmartContract: string): void {
        this.setState({
            activeSmartContract: selectedSmartContract
        });
    }

    public render(): any {
        return (
            <div className='page-container bx--grid' data-test-id='txn-page'>
                <div className='inner-container bx--row'>
                    <Sidebar/>
                    <div className='page-contents bx--col'>
                        <div className='titles-container'>
                            <span>{this.state.activeSmartContract} home</span>
                            <h1>Transactions</h1>
                        </div>
                        <div className='contents-container bx--row'>
                            <div className='contents-left bx--col'>
                                <p>Welcome to your transaction web view for <span className='disabled-smart-contract'>{this.state.activeSmartContract}</span>; use this to submit and evaluate new and existing transactions.</p>
                                <p>Unsure of where to start? Follow our introductory Getting Started section opposite.</p>
                                <br/>
                                <h5>Smart Contracts</h5>
                                <p>To switch to the web view for a different smart contract, select the smart contract below</p>
                                {this.getSmartContractItems(this.state.smartContracts)}
                            </div>
                            <div className='contents-right bx--col' id='contents-right'>
                                <h5>Getting Started</h5>
                                <Accordion>
                                    <AccordionItem title={'Create a new transaction'}></AccordionItem>
                                </Accordion>
                                <Accordion>
                                    <AccordionItem title={'Edit existing transactions'}></AccordionItem>
                                </Accordion>
                                <Accordion>
                                    <AccordionItem title={'Favouriting & deleting transactions'}></AccordionItem>
                                </Accordion>
                                <Accordion>
                                    <AccordionItem title={'Submitting and evaluating: what\'s the difference?'}></AccordionItem>
                                </Accordion>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }

    private getSmartContractItems(smartContracts: string[]): any {
        const smartContractItems: any = [];

        for (const contract of smartContracts) {
            if (contract === this.state.activeSmartContract) {
                smartContractItems.push(<li className='smart-contract-item disabled-smart-contract' key={contract}>{contract}</li>);
            } else {
                smartContractItems.push(<li className='smart-contract-item clickable-smart-contract' key={contract}  onClick={(): void => this.switchSmartContract(contract)}>{contract}</li>);
            }
        }

        return <ul>{smartContractItems}</ul>;
    }
}

export default TransactionHome;
