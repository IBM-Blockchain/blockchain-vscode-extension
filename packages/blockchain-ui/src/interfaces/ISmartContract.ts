import ITransaction from './ITransaction';

interface ISmartContract {
    name: string;
    version: string;
    channel: string;
    label: string;
    transactions: Array<ITransaction>;
    namespace: string;
    peerNames: string[];
}

export default ISmartContract;
