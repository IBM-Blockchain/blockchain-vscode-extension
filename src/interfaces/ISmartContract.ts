import ITransaction from './ITransaction';

interface ISmartContract {
    name: string;
    version: string;
    channel: string;
    label: string;
    transactions: Array<ITransaction>;
    namespace: string;
}

export default ISmartContract;
