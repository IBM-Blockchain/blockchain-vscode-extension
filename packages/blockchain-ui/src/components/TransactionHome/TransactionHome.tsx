import React, { Component } from 'react';
import './TransactionHome.scss';
import { SelectItem } from 'carbon-components-react';
import InlineSelect from '../InlineSelect/InlineSelect';
import TransactionTable from '../TransactionTable/TransactionTable';
import ISmartContract from '../../interfaces/ISmartContract';

interface HomeProps {
    gatewayName: string;
    activeSmartContract: ISmartContract;
    smartContracts: Array<ISmartContract>;
    switchSmartContract: (newActiveContract: string) => void;
    postMessageHandler: (command: string, data?: any) => void;
}

interface HomeState {
    gatewayName: string;
    activeSmartContractLabel: string;
    smartContractLabels: Array<string>;
    switchSmartContract: (newActiveContract: string) => void;
    postMessageHandler: (command: string, data?: any) => void;
}

class TransactionHome extends Component<HomeProps, HomeState> {

    constructor(props: Readonly<HomeProps>) {
        super(props);
        this.state = this.setUpState(this.props);
        this.smartContractSelectChangeHandler = this.smartContractSelectChangeHandler.bind(this);
    }

    componentDidUpdate(prevProps: HomeProps): void {
        if (prevProps.activeSmartContract.label !== this.props.activeSmartContract.label) {
            this.setState({
                activeSmartContractLabel: this.props.activeSmartContract.label
            });
        }
    }

    setUpState(receivedProps: HomeProps): HomeState {
        const contractLabels: Array<string> = [];
        for (const contract of receivedProps.smartContracts) {
            contractLabels.push(contract.label);
        }

        return {
            gatewayName: receivedProps.gatewayName,
            activeSmartContractLabel: receivedProps.activeSmartContract.label,
            smartContractLabels: contractLabels,
            switchSmartContract: receivedProps.switchSmartContract,
            postMessageHandler: receivedProps.postMessageHandler
        };
    }

    smartContractSelectChangeHandler(event: React.FormEvent<HTMLSelectElement>): void {
        const newSmartContractLabel: string = event.currentTarget.value;
        this.state.switchSmartContract(newSmartContractLabel);
    }

    populateSmartContractSelect(): Array<JSX.Element> {
        const options: Array<JSX.Element> = [];
        options.push(<SelectItem disabled={false} hidden={true} text={this.state.activeSmartContractLabel} value={this.state.activeSmartContractLabel}/>);
        for (const contract of this.state.smartContractLabels) {
            options.push(<SelectItem disabled={false} hidden={false} text={contract} value={contract}/>);
        }
        return options;
    }

    getRecentTransactions(): Array<any> {
        return [
            {
                id: '1',
                name: 'createAsset',
                arguments: 'AssetID: myAsset01',
                timestamp: '1:15pm',
                result: '✅'
            },
            {
                id: '2',
                name: 'readAsset',
                arguments: 'AssetID: myAsset01',
                timestamp: '1:32pm',
                result: '✅'
            },
            {
                id: '3',
                name: 'updateAsset',
                arguments: 'AssetID: myAsset01',
                timestamp: '1:42pm',
                result: '✅'
            },
            {
                id: '4',
                name: 'deleteAsset',
                arguments: 'AssetID: myAsset01',
                timestamp: '1:59pm',
                result: '✅'
            }
        ];
    }

    render(): JSX.Element {
        const tableRows: Array<any> = this.getRecentTransactions();

        return (
            <div className='page-container bx--grid' data-test-id='txn-page'>
                <div className='inner-container'>
                    <div className='page-contents'>
                        <div className='titles-container'>
                            <InlineSelect id='smart-contract-select' labelText='Smart contract' contents={this.populateSmartContractSelect()} onChangeCallback={this.smartContractSelectChangeHandler}/>
                            <h2>Transactions view for {this.state.gatewayName}</h2>
                        </div>
                        <div className='contents-container bx--row'>
                            <p>Welcome to the transaction webview for {this.state.activeSmartContractLabel}. View, submit and evaluate transactions, as well as import and export them. Navigate to alternative smart contracts under {this.state.gatewayName} through the top left.</p>
                            <br/>
                            <TransactionTable
                                id={'recent-txns-table'}
                                title={'Recent Transactions'}
                                description={'View your recent transactions, unused transactions will clear every 50 submitted.'}
                                rows={tableRows}
                                buttonId={'recent-txns-table-btn'}
                                buttonText={'Create a new transaction'}
                                buttonFunction={(): void => this.state.postMessageHandler('create')}
                            />
                        </div>
                        <br></br>
                        <div className='contents-container bx--row'>
                            <TransactionTable
                                id={'saved-txns-table'}
                                title={'Saved templates'}
                                description={'View your saved templates'}
                                rows={[{id: '1', name: 'No transaction templates saved'}]}
                                buttonId={'saved-txns-table-btn'}
                                buttonText={'Import a transaction'}
                                buttonFunction={(): void => this.state.postMessageHandler('import')}
                            />
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

export default TransactionHome;
