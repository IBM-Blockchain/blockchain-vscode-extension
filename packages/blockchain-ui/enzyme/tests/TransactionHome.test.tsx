// tslint:disable no-unused-expression
import React from 'react';
import { mount, shallow } from 'enzyme';
import TransactionHome from '../../src/components/TransactionHome/TransactionHome';
import ITransaction from '../../src/interfaces/ITransaction';
import ISmartContract from '../../src/interfaces/ISmartContract';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
chai.should();
chai.use(sinonChai);

describe('TransactionHome component', () => {

    let mySandBox: sinon.SinonSandbox;
    let switchSmartContractStub: sinon.SinonStub;
    let postMessageHandlerStub: sinon.SinonStub;

    const mockTxn: ITransaction = {
        name: 'mockTxn',
        parameters: [{
            description: '',
            name: 'name',
            schema: {}
        }],
        returns: {
            type: ''
        },
        tag: ['submit']
    };

    const greenContract: ISmartContract = {
        name: 'greenContract',
        version: '0.0.1',
        channel: 'mychannel',
        label: 'greenContract@0.0.1',
        transactions: [mockTxn],
        namespace: 'GreenContract'
    };

    const blueContract: ISmartContract = {
        name: 'blueContract',
        version: '0.0.1',
        channel: 'mychannel',
        label: 'blueContract@0.0.1',
        transactions: [mockTxn],
        namespace: 'BlueContract'
    };

    const mockState: {gatewayName: 'myGateway', smartContracts: Array<ISmartContract>, activeSmartContract: ISmartContract} = {
        gatewayName: 'myGateway',
        smartContracts: [greenContract, blueContract],
        activeSmartContract: greenContract
    };

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();
        switchSmartContractStub = mySandBox.stub().resolves();
        postMessageHandlerStub = mySandBox.stub().resolves();
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    it('should render the expected snapshot', async () => {
        const component: any = shallow(<TransactionHome {...mockState} switchSmartContract={switchSmartContractStub} postMessageHandler={postMessageHandlerStub}/>);
        expect(component.getElements()).toMatchSnapshot();
    });

    it('should correctly set the state', async () => {
        const component: any = mount(<TransactionHome {...mockState} switchSmartContract={switchSmartContractStub} postMessageHandler={postMessageHandlerStub}/>);
        component.state().should.deep.equal({
            activeSmartContractLabel: mockState.activeSmartContract.label,
            gatewayName: mockState.gatewayName,
            postMessageHandler: postMessageHandlerStub,
            smartContractLabels: [
                greenContract.label,
                blueContract.label
            ],
            switchSmartContract: switchSmartContractStub
        });
    });

    it('should attempt to switch the active smart contract if another contract is selected', async () => {
        const component: any = mount(<TransactionHome {...mockState} switchSmartContract={switchSmartContractStub} postMessageHandler={postMessageHandlerStub}/>);
        component.find('select').at(0).prop('onChange')( { currentTarget: { value: blueContract.label } } );
        switchSmartContractStub.should.have.been.calledOnceWithExactly(blueContract.label);
    });

    it('should update the active smart contract when a new one is passed down through props', async () => {
        const componentDidUpdateSpy: sinon.SinonSpy = mySandBox.spy(TransactionHome.prototype, 'componentDidUpdate');
        const component: any = mount(<TransactionHome {...mockState} switchSmartContract={switchSmartContractStub} postMessageHandler={postMessageHandlerStub}/>);
        component.state().activeSmartContractLabel.should.equal(greenContract.label);

        component.setProps({
            gatewayName: 'myGateway',
            activeSmartContract: blueContract,
            smartContracts: greenContract, blueContract
        });
        componentDidUpdateSpy.should.have.been.called;
        component.state().activeSmartContractLabel.should.equal(blueContract.label);
    });

    it('called the expected command when the recent transactions table button is clicked', async () => {
        const component: any = mount(<TransactionHome {...mockState} switchSmartContract={switchSmartContractStub} postMessageHandler={postMessageHandlerStub}/>);
        component.find('#recent-txns-table-btn').at(1).simulate('click');
        postMessageHandlerStub.should.have.been.calledWith('create');
    });

    it('called the expected command when the saved transactions table button is clicked', async () => {
        const component: any = mount(<TransactionHome {...mockState} switchSmartContract={switchSmartContractStub} postMessageHandler={postMessageHandlerStub}/>);
        component.find('#saved-txns-table-btn').at(1).simulate('click');
        postMessageHandlerStub.should.have.been.calledWith('import');
    });

});
