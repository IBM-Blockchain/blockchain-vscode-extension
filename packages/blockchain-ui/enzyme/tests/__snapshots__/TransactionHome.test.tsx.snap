// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionHome component should render the expected snapshot 1`] = `
Array [
  <div
    className="page-container bx--grid"
    data-test-id="txn-page"
  >
    <div
      className="inner-container"
    >
      <div
        className="page-contents"
      >
        <div
          className="titles-container"
        >
          <InlineSelect
            contents={
              Array [
                <SelectItem
                  disabled={false}
                  hidden={true}
                  text="greenContract@0.0.1"
                  value="greenContract@0.0.1"
                />,
                <SelectItem
                  disabled={false}
                  hidden={false}
                  text="greenContract@0.0.1"
                  value="greenContract@0.0.1"
                />,
                <SelectItem
                  disabled={false}
                  hidden={false}
                  text="blueContract@0.0.1"
                  value="blueContract@0.0.1"
                />,
              ]
            }
            id="smart-contract-select"
            labelText="Smart contract"
            onChangeCallback={[Function]}
          />
          <h2>
            Transactions view for 
            myGateway
          </h2>
        </div>
        <div
          className="contents-container bx--row"
        >
          <p>
            Welcome to the transaction webview for 
            greenContract@0.0.1
            . View, submit and evaluate transactions, as well as import and export them. Navigate to alternative smart contracts under 
            myGateway
             through the top left.
          </p>
          <br />
          <TransactionTable
            buttonFunction={[Function]}
            buttonId="recent-txns-table-btn"
            buttonText="Create a new transaction"
            description="View your recent transactions, unused transactions will clear every 50 submitted."
            id="recent-txns-table"
            rows={
              Array [
                Object {
                  "arguments": "AssetID: myAsset01",
                  "id": "1",
                  "name": "createAsset",
                  "result": "✅",
                  "timestamp": "1:15pm",
                },
                Object {
                  "arguments": "AssetID: myAsset01",
                  "id": "2",
                  "name": "readAsset",
                  "result": "✅",
                  "timestamp": "1:32pm",
                },
                Object {
                  "arguments": "AssetID: myAsset01",
                  "id": "3",
                  "name": "updateAsset",
                  "result": "✅",
                  "timestamp": "1:42pm",
                },
                Object {
                  "arguments": "AssetID: myAsset01",
                  "id": "4",
                  "name": "deleteAsset",
                  "result": "✅",
                  "timestamp": "1:59pm",
                },
              ]
            }
            title="Recent Transactions"
          />
        </div>
        <br />
        <div
          className="contents-container bx--row"
        >
          <TransactionTable
            buttonFunction={[Function]}
            buttonId="saved-txns-table-btn"
            buttonText="Import a transaction"
            description="View your saved templates"
            id="saved-txns-table"
            rows={
              Array [
                Object {
                  "id": "1",
                  "name": "No transaction templates saved",
                },
              ]
            }
            title="Saved templates"
          />
        </div>
      </div>
    </div>
  </div>,
]
`;
